code/string_algorithms/src/finite_automata/C/dfa.c

```diff
@@ -0,0 +1,349 @@
+/*
+ * Part of Cosmos by OpenGenus.
+ * Author : ABDOUS Kamel
+ */
+
+#include "dfa.h"
+#include <stdlib.h>
+
+/* This array is used when doing a dfs */
+static dfs_color_t* dfs_colors;
+
+/* Tells if a state is accessible, when removing useless states */
+static bool* access_bits;
+
+/* Stores the new ids of states, when removing useless states */
+static state_id_t* states_new_ids;
+
+/* Stores the number of coaccessible states of a given DFA */
+static int coaccess_states_nb;
+
+/*
+ * Depth-first search subroutine for accessible states.
+ */
+static void
+not_accessible_subroutine(dfa_t* dfa, state_id_t root, state_link_t* access_list);
+
+/*
+ * Reverse depth-first search subroutine for coaccessible states.
+ */
+static state_link_t*
+not_coaccessible_subroutine(dfa_t* dfa, state_id_t root, state_link_t* coaccess_list);
+
+/*
+ * Allocate necessary data structure for the given dfa.
+ * States ID's are initialized.
+ * Initialize all transitions to -1.
+ * @return 0 on success, -1 on failure.
+ */
+int
+allocate_dfa(int states_nb, dfa_t* dfa)
+{
+  dfa->alpha_size = 256;
+  dfa->states_nb = states_nb;
+
+  dfa->states = malloc(sizeof(state_t) * states_nb);
+  if (dfa->states == NULL)
+    return (-1);
+
+  dfa->transitions = malloc(sizeof(state_id_t*) * states_nb);
+  if (dfa->transitions == NULL) {
+    free(dfa->states);
+    return (-1);
+  }
+
+  int i, j;
```

Comment:
Move variables close to where they are used; `j` should be near its loop.

Reply:
If I do so, I'd declare `j` under the loop.